# A1_노선호_통계2일차

In [1]:
import numpy as np
from scipy import stats
import scipy.stats
import pandas as pd
import statsmodels.api as sm
import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns
from statsmodels.stats.proportion import proportions_ztest
%matplotlib inline

# [실습] weibull 분포

In [2]:
x = 2500
alpha= 5.2
beta = 2200
prob = stats.weibull_min.cdf(x,alpha,scale=beta)
print("P(X<=x): {:.3f}".format(1-prob))

P(X<=x): 0.143


### 이 제품이 2,500시간 이상 작동할 확률은14.3%이다!

# [실습] 이항분포 

In [3]:
n =5
for i in range(n+1):
    prob = stats.binom.pmf(k=i,n=n,p=0.3)
    print("P(X={0})= {1:.3f}".format(i,prob))

P(X=0)= 0.168
P(X=1)= 0.360
P(X=2)= 0.309
P(X=3)= 0.132
P(X=4)= 0.028
P(X=5)= 0.002


### 불량대수가 0일 확률은 16.8%이고
### 불량대수가 1일 확률은 36.0%이고
### 불량대수가 2일 확률은 30.9%이고
### 불량대수가 3일 확률은 13.2%이고
### 불량대수가 4일 확률은 2.8% 이고
### 불량대수가 5일 확률은 0.2%이다!

# [실습] 포아송 분포

In [4]:
mu = 2
prob = stats.poisson.pmf(3,mu)
cdf_prob = stats.poisson.cdf(2,mu)
print("1분당 {0}번의 전화가 걸려올 확률: {1:.1f}%".format(3,prob*100))
print("1분당 최대 {0}번의 전화가 걸려올 확률: {1:.1f}%".format(2,cdf_prob*100))

1분당 3번의 전화가 걸려올 확률: 18.0%
1분당 최대 2번의 전화가 걸려올 확률: 67.7%


In [5]:
mu = 3
prob = stats.poisson.pmf(2,mu)
# 2명의 손님이 도착
cdf_prob = stats.poisson.cdf(1,mu)
# 1명 이하의 손님이 도착
cdf_prob3 = stats.poisson.cdf(3,mu)
print("1시간에 {0}명의 손님이 도착할 확률: {1:.1f}%".format(2,prob*100))
print("1시간에 {0}명 이하의 손님이 도착할 확률: {1:.1f}%".format(1,cdf_prob*100))
print("1시간에 {0}명 이상의 손님이 도착할 확률: {1:.1f}%".format(3,(1-cdf_prob3)*100))

1시간에 2명의 손님이 도착할 확률: 22.4%
1시간에 1명 이하의 손님이 도착할 확률: 19.9%
1시간에 3명 이상의 손님이 도착할 확률: 35.3%


# [실습] 정규성 검정

In [6]:
from scipy.stats import shapiro

ds_2sample_t_test = pd.read_csv("/home/piai/Downloads/week2/2sample_t_test.csv",engine="python")
ds_2sample_t_test.head()

Mat_A  Mat_B  Total  Sub
0   73.4   68.7   73.4    1
1   77.0   71.4   77.0    1
2   73.7   69.8   73.7    1
3   73.3   75.3   73.3    1
4   73.1   71.3   73.1    1

In [7]:
ds_2sample_t_test=ds_2sample_t_test.dropna() # 데이터 전처리 
data1 = ds_2sample_t_test['Mat_A']
print(shapiro(data1))


ShapiroResult(statistic=0.9062636494636536, pvalue=0.2562994360923767)


In [8]:
statistic,p =stats.shapiro(data1)
print("Shapiro-Wilk Test: statistic={}, p-value={}".format(statistic,p))

Shapiro-Wilk Test: statistic=0.9062636494636536, p-value=0.2562994360923767


In [9]:
data2 = ds_2sample_t_test['Mat_B']
print(shapiro(data2))

ShapiroResult(statistic=0.9729753732681274, pvalue=0.9169777035713196)


In [10]:
statistic,p =stats.shapiro(data2)
print("Shapiro-Wilk Test: statistic={}, p-value={}".format(statistic,p))

Shapiro-Wilk Test: statistic=0.9729753732681274, p-value=0.9169777035713196


### "정규성 검정결과(95%신뢰 수준 사용시) p-value가 0.25..이므로 0.05보다 커서 정규분포를 따른다고 판단됨"

# [실습] 모평균 추정(모표준편차 아는 경우)

In [11]:
df =pd.DataFrame({"sample": [18,18,20,21,20,23,19,18,17,21,22,20,20,21,20,19,19,18,17,19]})
df.head()
lower,upper = stats.norm.interval(0.95, loc= np.mean(df), scale = 3.8/np.sqrt(20))
print("신뢰구간: ({0},{1})".format(lower.round(2),upper.round(2)))

신뢰구간: ([17.83],[21.17])


# [실습] 모평균추정(모표준편차 모르는 경우) 

In [12]:
df=pd.DataFrame({"sample":[73,71,74,69,70,73,70,68,75,72,70,72,73,70,70,72,71,70,75,72]})

lower,upper=stats.t.interval(0.95,len(df)-1,loc=np.mean(df), scale= scipy.stats.sem(df))
print("신뢰구간: ({0},{1})".format(lower.round(2),upper.round(2)))

신뢰구간: ([70.6],[72.4])


# [실습] 모분산 신뢰구간 추정

In [13]:
df= pd.DataFrame({226,228,226,225,232,228,227,229,225,230})

In [14]:
df.std()

0    2.410295
dtype: float64

In [15]:
df= pd.DataFrame({12,13,16,9,5,19,13,8,11,17,9,5,12,11,8,6,10,12,17,11,10,9,12,14,9,6,8,14,5,8,5,8,4,9,6,8,13,16,9,5,16,13,14,17,11,9,19,16,12,8})

In [16]:
df.std()

0    4.680839
dtype: float64

### 표준편차를 std() 함수로 소숫점 두자리까지 구하면 4.68 이 나온다.

In [17]:
a= (49)*(4.68)*(4.68)/(71.42)
b= (49)*(4.68)*(4.68)/(32.36)
print('(49)×(4.68)²÷(71.42) <= σ <=(49)×(4.68)²÷(32.36) ')

print(' 모분산의 95% 신뢰구간 = ({0:0.1f},{1:0.1f})'.format(a,b))

(49)×(4.68)²÷(71.42) <= σ <=(49)×(4.68)²÷(32.36) 
 모분산의 95% 신뢰구간 = (15.0,33.2)


# [실습] 모비율 신뢰구간 추정 

In [18]:
import math
a= 0.82-1.96*math.sqrt(0.82*(1 - 0.82)/50)
b= 0.82+1.96*math.sqrt(0.82*(1 - 0.82)/50)
print('0.82-1.96√(0.82×(1−0.82)÷50) < p <(0.82+1.96√(0.82×(1−0.82)÷50) ')
print()
print(' 모분산의 95% 신뢰구간 = ({0:0.2f},{1:0.2f})'.format(a,b))

0.82-1.96√(0.82×(1−0.82)÷50) < p <(0.82+1.96√(0.82×(1−0.82)÷50) 

 모분산의 95% 신뢰구간 = (0.71,0.93)


# [실습] 평균 검정(모집단 표준편차를 모르는 경우)

In [19]:
df = pd.DataFrame({'sample':[85.0,79.0,79.1,79.9,81.6,78.6,85.4,83.4,78.1,79.2]})

In [20]:
t_result = stats.ttest_1samp(df,78)

In [21]:
t,p=t_result.statistic.round(3), t_result.pvalue.round(3)

In [22]:
print("1-Sample t-test")
print("t 통계량: {}".format(t))
print("p-value: {}".format(p))


1-Sample t-test
t 통계량: [3.379]
p-value: [0.008]


### "유의수준 5%에서 검정결과 P값이 0.008이므로 개선활동이 고객만족도 변화를 시켰다고 할 수 없다."

# [실습] 2 Sample t-test

In [23]:
df1= pd.DataFrame({'sample':[1.883,1.715,1.799,1.768,1.711,1.832,1.427,1.344]})
df2= pd.DataFrame({'sample':[1.435,1.572,1.486,1.511,1.457,1.548,1.404,1.883]})
t_result = stats.ttest_ind(df1,df2)
t,p = t_result.statistic.round(3), t_result.pvalue.round(3)
print(" 2-Sample t-test")
print(" t:{}".format(t))
print("p:{}".format(p))

 2-Sample t-test
 t:[1.7]
p:[0.111]


### 유의수준 5%에서 검정결과 P값이 0.111이므로 0.05를 넘어서 고객간에 고객만족도의 차이가 없다." 

# [실습] Paired t-test

In [24]:
df1= pd.DataFrame({'before':[720,589,780,648,720,589,780,648,780,648]})
df2= pd.DataFrame({'After':[710,580,787,712,750,600,782,670,790,680]})

t_result = stats.ttest_rel(df1,df2)
t, p = t_result.statistic.round(3), t_result.pvalue.round(3)
print(" Paired t-test")
print("t: {}".format(t))
print("p:{}".format(p))


 Paired t-test
t: [-2.266]
p:[0.05]


### "유의수준 5%에서 검정결과 p값이 0.05이므로  학원 수강 전후의 점수는 차이가 있다고 할 수 있다. "

# [실습] 1 Proportion test 

In [25]:
from statsmodels.stats.proportion import proportions_ztest

In [26]:
count = 40
n_obs=100
value = 0.5
stat , pval = proportions_ztest(count,n_obs,value)
print(" 1 Proportion test")
print(" z:{0:0.3f}".format(stat))
print("p:{0:0.3f}".format(pval))


 1 Proportion test
 z:-2.041
p:0.041


### "p가 0.041 이므로 0.05보다 작기 때문에 h1 채택하고 차이가 있다."

# [실습]2 Proportion test

In [27]:
count = np.array([14,5])
nobs= np. array([1200,1200])

stat,pval = proportions_ztest(count,nobs)
print(" 2 Proprotion test")
print("z:{0:0.3f}".format(stat))
print("p:{0:0.3f}".format(pval))



 2 Proprotion test
z:2.073
p:0.038


### "p-value <0.05 이므로 두 공장의 불량률은 차이가 있다고 할 수 없다. "